Src/VimCore/MotionUtil.fs

```diff
@@ -348,6 +348,210 @@ type MatchingTokenKind =
 
     | Braces
 
+type internal BlockUtil() =
+
+    /// Get the block span for the specified char at the given context point
+    member x.GetBlock (blockKind: BlockKind) contextPoint =
+
+        // Blocks in vim are a messy combination of heuristics and
+        // a lot of special cases. The main complication is the
+        // handling of parentheses in string and character literals.
+        // The idea is to match parentheses the same way that the
+        // programming language would, that is if they are syntactic
+        // elements, match them that way, and if they are string
+        // elements, match them that way instead.
+
+        // If the context point for a block is outside a string literal,
+        // we don't want unbalanced parentheses in strings throwing us
+        // off. Likewise, if we are inside a string literal, we don't
+        // want a parenthesis inside the string to match a syntactic
+        // parenthesis outside the string. So if possible we want
+        // to allow valid blocks in, say, a regular expression in
+        // a string, as well as valid syntactic expression blocks.
+
+        // Likewise, unbalances parentheses or stray quote marks in
+        // comments can also be problematic, but vim does very little
+        // to accomodate those cases.
+
+        // In any case, this is an inexact science because different
+        // programming languages have different ways of representing
+        // and encoding strings. The goal is to do the intuitive thing
+        // from a programming point of view in as many cases as
+        // possible.
+
+        // Our "univeral string" (covering common cases in C, C++, C#,
+        // F#, Python and Java, etc.) is text surrounded with single
+        // or double quotes and using backslash to escape the quote
+        // character.
+
+        // A more robust solution would be to use an abstract syntax
+        // tree supplied by a language service appropriate for the
+        // current buffer but even that approach has drawbacks. The
+        // current approach works as well for a programming snippet
+        // in a text file as it does in an actual source file.
+
+        let startChar, endChar = blockKind.Characters
+
+        // Is the char at the given point escaped?
+        let isEscaped point =
+            match SnapshotPointUtil.TrySubtractOne point with
+            | None -> false
+            | Some point -> SnapshotPointUtil.GetChar point = '\\'
+
+        // Is the char at the given point double escaped?
+        let isDoubleEscaped point =
+            match SnapshotPointUtil.TrySubtract point 2, SnapshotPointUtil.TrySubtract point 2 with
+            | None, _

```

Comment:
Huh, didn't realize F# had this type of fall through behavior.

Reply:
The underscore matches a tuple. In any case, I didn't change that code; I just had to move it to a new type for reuse. So you probably wrote the original code!